### ISSUE TITLE
invulnerableDimensionChange -> wasMovedByPortal

### ISSUE THREAD

**Issue by Runemoro** [ModCoderPack/MCPBot-Issues#624]
- [ ] `field_184851_cj` in `net.minecraft.entity.player.EntityPlayerMP`: `invulnerableDimensionChange` -> `wasMovedByPortal`

Although the only place `invulnerableDimensionChange` is set to true is in the `EntityPlayerMP.changeDimension` method, its function is to disable anti-cheat checks ("Player ___ moved wrongly!") for the rest of the tick after the player has collided with a portal block (the only two vanilla portals being the nether and end portals, both of which call `changeDimension`). It is not called by commands because anti-cheat checks don't run after those.

Therefore, it shouldn't be called `invulnerable`, because it does more than just making the player invulnerable for one tick (probably as a workaround for [MC-98153](https://bugs.mojang.com/browse/MC-98153)), and it shouldn't be called `DimensionChange` since the only reason it's not set in other cases is that Minecraft hasn't yet implemented portal blocks that don't change your dimension.

**Comment by Pokechu22** [comment]
> Minecraft hasn't yet implemented portal blocks that don't change your dimension.

Minecraft does have those: the [end gateway](https://minecraft.gamepedia.com/End_Gateway_(block)).  They simply call `setPositionAndUpdate` when an entity enters them.  I'm pretty sure that this is only intended for use in dimension changes.

**Comment by Runemoro** [comment]
Yes, but it's the tile entity's update method that handles the does the teleport for end gateways:

```java
public void update() {
    // ...
    if (!this.world.isRemote) {
        List<Entity> list = this.world.<Entity>getEntitiesWithinAABB(Entity.class, new AxisAlignedBB(this.getPos()));

        if (!list.isEmpty()) {
            this.teleportEntity(list.get(0));
        }
        // ...
    }
    // ...
}
```

If it was the block collision that triggered the teleport, then `invulnerableDimensionChange` would need to be set to true to disable anti-cheat for that tick. Otherwise, the player's position would get reverted:

```java
this.player.move(MoverType.PLAYER, d7, d8, d9);
this.player.onGround = packetIn.isOnGround();
double d12 = d8;
d7 = d4 - this.player.posX;
d8 = d5 - this.player.posY;

if (d8 > -0.5D || d8 < 0.5D) {
    d8 = 0.0D;
}

d9 = d6 - this.player.posZ;
d11 = d7 * d7 + d8 * d8 + d9 * d9;
boolean flag = false;

if (!this.player.isInvulnerableDimensionChange() && d11 > 0.0625D && !this.player.isPlayerSleeping() && !this.player.interactionManager.isCreative() && this.player.interactionManager.getGameType() != GameType.SPECTATOR) {
    flag = true;
    LOGGER.warn("{} moved wrongly!", (Object)this.player.getName());
}
```
